python/cuml/internals/base.pyx

```diff
@@ -423,6 +478,132 @@ def _determine_stateless_output_type(output_type, input_obj):
 
     # If we are using 'input', determine the the type from the input object
     if temp_output == 'input':
-        temp_output = cuml.common.input_utils.determine_array_type(input_obj)
+        temp_output = cuml.internals.input_utils.determine_array_type(input_obj)
 
     return temp_output
+
+
+class UniversalBase(Base):
+
+    def dispatch_func(self, func_name, *args, **kwargs):
+        """
+        This function will dispatch calls to training and inference according
+        to the global configuration. It should work for all estimators
+        sufficiently close the scikit-learn implementation as it uses
+        it for training and inferences on host.
+
+        Parameters
+        ----------
+        func_name : string
+            name of the function to be dispatched
+        args : arguments
+            arguments to be passed to the function for the call
+        kwargs : keyword arguments
+            keyword arguments to be passed to the function for the call
+        """
+        # look for current device_type
+        device_type = cuml.global_settings.device_type
+        if device_type == DeviceType.device:
+            # call the original cuml method
+            cuml_func_name = '_' + func_name
+            if hasattr(self, cuml_func_name):
+                cuml_func = getattr(self, cuml_func_name)
+                return cuml_func(*args, **kwargs)
+            else:
+                raise ValueError('Function "{}" could not be found in'
+                                 ' the cuML estimator'.format(cuml_func_name))
+
+        elif device_type == DeviceType.host:
```

Comment:
This branch could use some modularization.

Reply:
Agreed. This is actually part of #4918 and updated in #5001, so I'm going to defer discussion to that second PR. CC @viclafargue

Reply:
Yes, we should definitely split this function. Sure, let's discuss this in #5001.